SpotLightShadow.cameraAutoUpdate / Light.shadowAutoUpdate

mbredif opened the pull request:
Preventing the Autoupdating of the `SpotLight.shadow.camera` was previously performed through the hack of [overwriting](https://github.com/mrdoob/three.js/compare/dev...mbredif:SpotLightCameraUpdateOptOut?expand=1#diff-3a405b9c95eb5dd79f92e5ba070fafa9L115) the SpotLightShadow with a LightShadow. 

Instead, this PR proposes a new property ~`SpotLightShadow.cameraAutoUpdate`~`Light.shadowAutoUpdate` that makes optional the fitting of the shadow camera frustum to the square cone defined by light.angle/light.distance.

(this PR is the item 2 of #14658, extracted here for an easier review)

== Edit ==
Instead of overwriting the shadow, an alternative hack to get the same result would be to set `this.shadow.isSpotLightShadow = false`. As that was the sole use of `isSpotLightShadow`, this PR is essentially moving and renaming `Light.shadow.isSpotLightShadow` to `Light.shadowAutoUpdate`.

mbredif committed:
SpotLightShadow.cameraAutoUpdate

WestLangley commented on src/lights/SpotLightShadow.js:
```diff
@@ -20,6 +21,8 @@ SpotLightShadow.prototype = Object.assign( Object.create( LightShadow.prototype
 
 	update: function ( light ) {
 
+		if ( ! this.cameraAutoUpdate ) return;
```
Instead, I would do the following in the calling method:
```javascript
if ( shadow.isSpotLightShadow && shadow.cameraAutoUpdate ) {

	shadow.update( light );

}
```

mbredif replied:
done

WestLangley commented:
I definitely support this.

I also support removal of the need for this pattern:
```javascript
light.shadow = new THREE.LightShadow( ... );
```

Forcing the shadow frustum to auto-size to the cone can have the side-effect of causing pixelated shadows.

mbredif committed:
more examples updated

mbredif committed:
review

mbredif commented:
2 more commits : 
- I found 2 more usages of the `light.shadow = new THREE.LightShadow( ... );` pattern in the examples.
- I moved the `cameraAutoUpdate` check to the calling site, as suggested in the review.

mbredif commented:
I am not sure about 2da8234 though, at it breaks http://localhost:8000/examples/webgl_lights_spotlight.html due to https://github.com/mrdoob/three.js/blob/2cbd28958be86cb5f34ee5532a3ee18a23467b94/examples/webgl_lights_spotlight.html#L142 which is now ungarded (it is required to have the spotlight and camera helpers in sync)

mbredif committed:
SpotLightShadow.cameraAutoUpdate -> Light.shadowAutoUpdate

mbredif commented:
@WestLangley what do you think of this new commit 64b191c ?
I moved `SpotLightShadow.cameraAutoUpdate` to `Light.shadowAutoUpdate`, updated the docs, and added the (conditional) light shadow update to the SpotLight helper. I also remove the now unused `isSpotLightShadow`.

WestLangley commented on docs/api/lights/SpotLight.html:
```diff
@@ -175,6 +175,13 @@ <h3>[property:Object3D target]</h3>
+		<h3>[property:Boolean shadowAutoUpdate]</h3>
+		<p>
+			If set to *true*, the shadow camera will be automatically updated each frame to match the
+			geometry of the spotLight (a square frustum tightly fitting the cone of the spot light defined
```
If set to *true*, the shadow camera frustum will be automatically sized to match the cone of the spotLight

mbredif replied:
done

WestLangley commented on docs/api/lights/SpotLight.html:
```diff
@@ -175,6 +175,13 @@ <h3>[property:Object3D target]</h3>
 			The spotlight will now track the target object.
 		</p>
 
+		<h3>[property:Boolean shadowAutoUpdate]</h3>
```
[property:Boolean autoSizeShadowFrustum] ?

WestLangley commented on docs/api/lights/shadows/LightShadow.html:
```diff
@@ -134,6 +134,11 @@ <h3>[method:Object toJSON]()</h3>
 
+		<h3>[method:void update]( [param:Light light] )</h3>
+		<p>
+		Updates this LightShadow (e.g. [page:SpotLightShadow.update SpotLightShadow.update]).
```
This method is akin to a virtual base class and does not do anything. Define what it does where it is overloaded.

WestLangley commented:
Actually, I think some others need to chime in regarding both the feature and the implementation details.

You are not "auto-updating" the shadow camera frustum, you are "auto-sizing" it. At least that is how I see it.

mbredif commented:
The naming shadowAutoUpdate refers to what it does for all lights (it guards a call to shadow.update, similarly to matrixAutoUpdate) rather than the previous  naming cameraAutoUpdate or autoSizeShadowFrustum which refers to what it does for spotlights. It is thus more generic (could be used unifiedly for using a custom shadow update function) and make isSpotLightShadow unnecessary.

mbredif committed:
review

mbredif commented:
Up !

There is no new functionality here. The goal is to provide a better API for the hack of overwriting the SpotLightShadow with a LightShadow to prevent the lightshadow update.

The alternatives are : 
A- renaming `LightShadow.isSpotLightShadow` to `LightShadow.autoUpdate`
B- moving `LightShadow.isSpotLightShadow` to `Light.shadowAutoUpdate`
C- use a more specific naming like `autoSizeShadowFrustum` or `cameraAutoUpdate`, but that opens the door for users to use the LightShadow.update function as a generic custom callback, guarded by a flag with a specific naming.

(Complementarily the `autoUpdate` approach, a `needsUpdate` approach could be implemented as well for a 1-time updating dirty flag.)

This PR implements B without the `needsUpdate` approach. What API would you prefer @WestLangley @mrdoob ?

Mugen87 commented:
Implemented via #19715.
